File: codes/go/array_every/array_every.go
package main

import (
	"encoding/json"
	"fmt"
	"reflect"
	"strings"
)

// type any interface{}
type array []any
type object map[string]any

func arrayReduce(callbackFunction func(any, any, int, array) any, anArray array, initialValue any) any {
    // JavaScript-like Array.reduce() function
    result := initialValue
    for arrayItemIndex, arrayItem := range anArray {
        result = callbackFunction(result, arrayItem, arrayItemIndex, anArray)
    }
    return result
}

func optionalChaining(anything any, objectPropertiesArray ...any) any {
    anythingType := reflect.TypeOf(anything).Kind()
    if (((anythingType != reflect.Map) && (anythingType != reflect.Slice)) || (len(objectPropertiesArray) == 0)) {
        return anything
    }
    return arrayReduce(func(currentResult any, currentItem any, _ int, _ array) any {
        if (currentResult == nil && (anythingType == reflect.Map) && (reflect.TypeOf(currentItem).Kind() == reflect.String)) {
            return anything.(object)[currentItem.(string)]
        }
        if (currentResult == nil && (anythingType == reflect.Slice) && (reflect.TypeOf(currentItem).Kind() == reflect.Int) && (currentItem.(int) >= 0) && (len(anything.(array)) > currentItem.(int))) {
            return anything.(array)[currentItem.(int)]
        }
        if (reflect.TypeOf(currentResult).Kind() == reflect.Map && (reflect.TypeOf(currentItem).Kind() == reflect.String)) {
            return currentResult.(object)[currentItem.(string)]
        }
        if ((reflect.TypeOf(currentResult).Kind() == reflect.Slice) && (reflect.TypeOf(currentItem).Kind() == reflect.Int) && (currentItem.(int) >= 0) && (len(currentResult.(array)) > currentItem.(int))) {
            return currentResult.(array)[currentItem.(int)]
        }
        return nil
    }, objectPropertiesArray, nil)
}

func ternary(trueCondition bool, callbackFunctionIfConditionTrue func() any, callbackFunctionIfConditionFalse func() any) any {
    if (trueCondition == true) {
        return callbackFunctionIfConditionTrue()
    }
    return callbackFunctionIfConditionFalse()
}

func jsonStringify(restArguments ...any) string {
	jsonStringifyInner := func(anythingInner any, prettyInner bool, indentInner string) string {
		if (prettyInner == true) {
			jsonStringifyInnerResult, err := json.MarshalIndent(anythingInner, "", indentInner)
			if (err == nil) {
				return string(jsonStringifyInnerResult)
			}
			return "null"
		}
        jsonStringifyInnerResult, err := json.Marshal(anythingInner)
		if (err == nil) {
			return strings.ReplaceAll(string(jsonStringifyInnerResult), ",", ", ")
		}
		return "null"
    }
    var anything any = restArguments[0]
	prettyDefault := false
	indentDefault := "    "
	var pretty any = prettyDefault
	var indent any = indentDefault
	if (len(restArguments) == 1) {
		return jsonStringifyInner(anything, prettyDefault, indentDefault)
	}
	var optionalArgument any = restArguments[1]
	if (optionalArgument == false) {
		return jsonStringifyInner(anything, prettyDefault, indentDefault)
	}
	if (reflect.TypeOf(optionalArgument).Kind() == reflect.Map) {
		pretty = optionalChaining(optionalArgument, "pretty")
		indent = optionalChaining(optionalArgument, "indent")
		pretty = ternary((pretty == nil), func() any { return prettyDefault }, func() any { return pretty })
		indent = ternary((indent == nil), func() any { return indentDefault }, func() any { return indent })
		return jsonStringifyInner(anything, pretty.(bool), indent.(string))
	}
	if (optionalArgument == true) {
		if (len(restArguments) >= 3) {
			var additionalArgument2 any = restArguments[2]
			if (reflect.TypeOf(additionalArgument2).Kind() == reflect.String) {
				indent = additionalArgument2
			}
		}
		pretty = optionalArgument
		return jsonStringifyInner(anything, pretty.(bool), indent.(string))
	}
	return jsonStringifyInner(anything, prettyDefault, indentDefault)
}

func sprint(restArguments ...any) {
    var newArray = []string{}
    for _, argument := range restArguments {
        argumentType := reflect.TypeOf(argument).Kind()
        if (argumentType == reflect.Slice && (len(argument.(array)) == 1)) {
            newArray = append(newArray, jsonStringify(argument.(array)[0]))
            continue
        }
        if (argumentType == reflect.String) {
			newArray = append(newArray, argument.(string))
            continue
        }
    }
    fmt.Println(strings.Join(newArray, ""))
}

func arrayEveryV1(callbackFunction func(any, int, array) bool, anArray array) bool {
    // JavaScript-like Array.every() function
    isConditionMatch := false
    for arrayItemIndex, arrayItem := range anArray {
        isConditionMatch = callbackFunction(arrayItem, arrayItemIndex, anArray)
        if (isConditionMatch == false) {
            break
        }
    }
    return isConditionMatch
}

func arrayEveryV2(callbackFunction func(any, int, array) bool, anArray array) bool {
    // JavaScript-like Array.every() function
    isConditionMatch := false
    for arrayItemIndex, arrayItem := range anArray {
        isConditionMatch = callbackFunction(arrayItem, arrayItemIndex, anArray)
        if (isConditionMatch == false) {
            return isConditionMatch
        }
    }
    return isConditionMatch
}

func arrayEveryV3(callbackFunction func(any, int, array) bool, anArray array) bool {
    // JavaScript-like Array.every() function
    for arrayItemIndex, arrayItem := range anArray {
        isConditionMatch := callbackFunction(arrayItem, arrayItemIndex, anArray)
        if (isConditionMatch == false) {
            return false
        }
    }
    return true
}

func arrayEveryV4(callbackFunction func(any, int, array) bool, anArray array) bool {
    // JavaScript-like Array.every() function
    for arrayItemIndex, arrayItem := range anArray {
        if (callbackFunction(arrayItem, arrayItemIndex, anArray) == false) {
            return false
        }
    }
    return true
}

func main() {
    fmt.Println("\n// JavaScript-like Array.every() in Go Slice")

    numbers := array{12, 34, 27, 23, 65, 93, 36, 87, 4, 254}
    sprint("numbers: ", jsonStringify(numbers))

    var isAllNumberLessThan500 bool
    var isAllNumberMoreThan500 bool

    fmt.Println("// using JavaScript-like Array.every() function \"arrayEveryV1\"")

    isAllNumberLessThan500 = arrayEveryV1(func(number any, _ int, _ array) bool {
        return (number.(int) < 500)
    }, numbers)
    sprint("is all number < 500: ", jsonStringify(isAllNumberLessThan500))
    // is all number < 500: true

    isAllNumberMoreThan500 = arrayEveryV1(func(number any, _ int, _ array) bool {
        return (number.(int) > 500)
    }, numbers)
    sprint("is all number > 500: ", jsonStringify(isAllNumberMoreThan500))
    // is all number > 500: false

    fmt.Println("// using JavaScript-like Array.every() function \"arrayEveryV2\"")

    isAllNumberLessThan500 = arrayEveryV2(func(number any, _ int, _ array) bool {
        return (number.(int) < 500)
    }, numbers)
    sprint("is all number < 500: ", jsonStringify(isAllNumberLessThan500))
    // is all number < 500: true

    isAllNumberMoreThan500 = arrayEveryV2(func(number any, _ int, _ array) bool {
        return (number.(int) > 500)
    }, numbers)
    sprint("is all number > 500: ", jsonStringify(isAllNumberMoreThan500))
    // is all number > 500: false

    fmt.Println("// using JavaScript-like Array.every() function \"arrayEveryV3\"")

    isAllNumberLessThan500 = arrayEveryV3(func(number any, _ int, _ array) bool {
        return (number.(int) < 500)
    }, numbers)
    sprint("is all number < 500: ", jsonStringify(isAllNumberLessThan500))
    // is all number < 500: true

    isAllNumberMoreThan500 = arrayEveryV3(func(number any, _ int, _ array) bool {
        return (number.(int) > 500)
    }, numbers)
    sprint("is all number > 500: ", jsonStringify(isAllNumberMoreThan500))
    // is all number > 500: false

    fmt.Println("// using JavaScript-like Array.every() function \"arrayEveryV4\"")

    isAllNumberLessThan500 = arrayEveryV4(func(number any, _ int, _ array) bool {
        return (number.(int) < 500)
    }, numbers)
    sprint("is all number < 500: ", jsonStringify(isAllNumberLessThan500))
    // is all number < 500: true

    isAllNumberMoreThan500 = arrayEveryV4(func(number any, _ int, _ array) bool {
        return (number.(int) > 500)
    }, numbers)
    sprint("is all number > 500: ", jsonStringify(isAllNumberMoreThan500))
    // is all number > 500: false

    fmt.Println("\n// JavaScript-like Array.every() in Go Slice of maps")

    products := array{
        object{
            "code": "pasta",
            "price": 321,
        },
        object{
            "code": "bubble_gum",
            "price": 233,
        },
        object{
            "code": "potato_chips",
            "price": 5,
        },
        object{
            "code": "towel",
            "price": 499,
        },
    }

    sprint("products: ", jsonStringify(products, object{"pretty": true}))

    var isAllProductPriceLessThan500 bool
    var isAllProductPriceMoreThan500 bool

    fmt.Println("// using JavaScript-like Array.every() function \"arrayEveryV1\"")

    isAllProductPriceLessThan500 = arrayEveryV1(func(product any, _ int, _ array) bool {
        return (product.(object)["price"].(int) < 500)
    }, products)
    sprint("is all product price < 500: ", jsonStringify(isAllProductPriceLessThan500))
    // is all product price < 500: true

    isAllProductPriceMoreThan500 = arrayEveryV1(func(product any, _ int, _ array) bool {
        return (product.(object)["price"].(int) > 500)
    }, products)
    sprint("is all product price > 500: ", jsonStringify(isAllProductPriceMoreThan500))
    // is all product price > 500: false

    fmt.Println("// using JavaScript-like Array.every() function \"arrayEveryV2\"")

    isAllProductPriceLessThan500 = arrayEveryV2(func(product any, _ int, _ array) bool {
        return (product.(object)["price"].(int) < 500)
    }, products)
    sprint("is all product price < 500: ", jsonStringify(isAllProductPriceLessThan500))
    // is all product price < 500: true

    isAllProductPriceMoreThan500 = arrayEveryV2(func(product any, _ int, _ array) bool {
        return (product.(object)["price"].(int) > 500)
    }, products)
    sprint("is all product price > 500: ", jsonStringify(isAllProductPriceMoreThan500))
    // is all product price > 500: false

    fmt.Println("// using JavaScript-like Array.every() function \"arrayEveryV3\"")

    isAllProductPriceLessThan500 = arrayEveryV3(func(product any, _ int, _ array) bool {
        return (product.(object)["price"].(int) < 500)
    }, products)
    sprint("is all product price < 500: ", jsonStringify(isAllProductPriceLessThan500))
    // is all product price < 500: true

    isAllProductPriceMoreThan500 = arrayEveryV3(func(product any, _ int, _ array) bool {
        return (product.(object)["price"].(int) > 500)
    }, products)
    sprint("is all product price > 500: ", jsonStringify(isAllProductPriceMoreThan500))
    // is all product price > 500: false

    fmt.Println("// using JavaScript-like Array.every() function \"arrayEveryV4\"")

    isAllProductPriceLessThan500 = arrayEveryV4(func(product any, _ int, _ array) bool {
        return (product.(object)["price"].(int) < 500)
    }, products)
    sprint("is all product price < 500: ", jsonStringify(isAllProductPriceLessThan500))
    // is all product price < 500: true

    isAllProductPriceMoreThan500 = arrayEveryV4(func(product any, _ int, _ array) bool {
        return (product.(object)["price"].(int) > 500)
    }, products)
    sprint("is all product price > 500: ", jsonStringify(isAllProductPriceMoreThan500))
    // is all product price > 500: false
}

File: codes/gnu_octave/array_every.m
function stringrepeatresult = stringrepeat(astring, count)
    result = "";
    for i = (1:1:count) % (start:step:stop)
        result = cstrcat(result, astring);
    end
    stringrepeatresult = result;
end

function arrayreduceresult = arrayreduce(callbackfunction, anarray, initialvalue)
    % JavaScript-like Array.reduce() function
    currentresult = initialvalue;
    for arrayitemindex = (1:1:numel(anarray)) % (start:step:stop)
        arrayitem = anarray{arrayitemindex};
        currentresult = callbackfunction(currentresult, arrayitem, arrayitemindex, anarray);
    end
    arrayreduceresult = currentresult;
end

function optionalchainingresult = optionalchaining(anything, varargin)
    % JavaScript-like Optional Chaining Operator (?.) function
    % varargin == objectpropertiesarray
    if ((~isstruct(anything) && ~iscell(anything)) || isempty(varargin))
        optionalchainingresult = anything;
        return;
    end
    function arrayreducecallbackresult = arrayreducecallback(currentresult, currentitem, varargin)
        % varargin == (arrayitemindex, anarray)
        if ((isempty(currentresult) && isstruct(anything) && (ischar(currentitem) || ischar(currentitem))) && isfield(anything, currentitem))
            arrayreducecallbackresult = anything.(currentitem);
            return;
        end
        if (isempty(currentresult) && iscell(anything) && isnumeric(currentitem) && (currentitem >= 1) && (numel(anything) > currentitem))
            arrayreducecallbackresult = anything{currentitem};
            return;
        end
        if ((isstruct(currentresult) && (ischar(currentitem) || ischar(currentitem))) && isfield(currentresult, currentitem))
            arrayreducecallbackresult = currentresult.(currentitem);
            return;
        end
        if (iscell(currentresult) && isnumeric(currentitem) && (currentitem >= 1) && (numel(currentresult) > currentitem))
            arrayreducecallbackresult = currentresult{currentitem};
            return;
        end
        arrayreducecallbackresult = {};
    end
    optionalchainingresult = arrayreduce(@arrayreducecallback, varargin, {});
end

function ternaryresult = ternary(truecondition, callbackfunctionifconditiontrue, callbackfunctionifconditionfalse)
    if (islogical(truecondition) && (truecondition == true))
        ternaryresult = callbackfunctionifconditiontrue();
        return;
    end
    ternaryresult = callbackfunctionifconditionfalse();
end

function jsonstringifyresult = jsonstringify(anything, varargin)
    indentlevel = 0;
    function jsonstringifyinnerresult = jsonstringifyinner(anythinginner, prettyinner, indentinner)
        if isempty(anythinginner)
            jsonstringifyinnerresult = "null";
            return;
        end
        if (ischar(anythinginner) || ischar(anythinginner))
            jsonstringifyinnerresult = cstrcat("""", anythinginner, """");
            return;
        end
        if isnumeric(anythinginner)
            jsonstringifyinnerresult = num2str(anythinginner);
            return;
        end
        if (islogical(anythinginner) && (anythinginner == true))
            jsonstringifyinnerresult = "true";
            return;
        end
        if (islogical(anythinginner) && (anythinginner == false))
            jsonstringifyinnerresult = "false";
            return;
        end
        if iscell(anythinginner)
            if isempty(anythinginner)
                jsonstringifyinnerresult = "[]";
                return;
            end
            indentlevel = indentlevel + 1;
            result = ternary((pretty == true), @() (cstrcat("[", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() ("["));
            for arrayitemindex = (1:1:numel(anythinginner)) % (start:step:stop)
                arrayitem = anythinginner{arrayitemindex};
                result = cstrcat(result, jsonstringifyinner(arrayitem, prettyinner, indentinner));
                if (arrayitemindex ~= numel(anythinginner))
                    result = ternary((pretty == true), @() (cstrcat(result, ",", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() (cstrcat(result, ", ")));
                end
            end
            indentlevel = indentlevel - 1;
            result = ternary((pretty == true), @() (cstrcat(result, sprintf("\n"), stringrepeat(indentinner, indentlevel), "]")), @() (cstrcat(result, "]")));
            jsonstringifyinnerresult = result;
            return;
        end
        if isstruct(anythinginner)
            if (numel(fieldnames(anythinginner)) == 0)
                jsonstringifyinnerresult = "{}";
                return;
            end
            indentlevel = indentlevel + 1;
            result = ternary((pretty == true), @() (cstrcat("{", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() ("{"));
            objectkeys = fieldnames(anythinginner);
            for objectentryindex = (1:1:numel(objectkeys)) % (start:step:stop)
                objectkey = objectkeys{objectentryindex};
                objectvalue = anythinginner.(objectkey);
                result = cstrcat(result, """", objectkey, """: ", jsonstringifyinner(objectvalue, prettyinner, indentinner));
                if (objectentryindex ~= numel(objectkeys))
                    result = ternary((pretty == true), @() (cstrcat(result, ",", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() (cstrcat(result, ", ")));
                end
            end
            indentlevel = indentlevel - 1;
            result = ternary((pretty == true), @() (cstrcat(result, sprintf("\n"), stringrepeat(indentinner, indentlevel), "}")), @() (cstrcat(result, "}")));
            jsonstringifyinnerresult = result;
            return;
        end
        jsonstringifyinnerresult = "null";
    end
    prettydefault = false;
    indentdefault = "    ";
    pretty = prettydefault;
    indent = indentdefault;
    if isempty(varargin)
        jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
        return;
    end
    optionalargument = varargin{1};
    if (islogical(optionalargument) && (optionalargument == false))
        jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
        return;
    end
    if isstruct(optionalargument)
        pretty = optionalchaining(optionalargument, "pretty");
        indent = optionalchaining(optionalargument, "indent");
        pretty = ternary(isempty(pretty), @() (prettydefault), @() (pretty));
        indent = ternary(isempty(indent), @() (indentdefault), @() (indent));
        jsonstringifyresult = jsonstringifyinner(anything, pretty, indent);
        return;
    end
    if (islogical(optionalargument) && (optionalargument == true))
        if (numel(varargin) >= 2)
            optionalargument2 = varargin{2};
            if (ischar(optionalargument2) || ischar(optionalargument2))
                indent = optionalargument2
            end
        end
        pretty = optionalargument;
        jsonstringifyresult = jsonstringifyinner(anything, pretty, indent);
        return;
    end
    jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
end

function sprint(varargin)
    currentresult = "";
    for argumentindex = (1:1:numel(varargin)) % (start:step:stop)
        argument = varargin{argumentindex};
        if (iscell(argument) && (numel(argument) == 1))
            argumentnew = jsonstringify(argument{1});
            argumentnew = strrep(argumentnew, """{", "{");
            argumentnew = strrep(argumentnew, """[", "[");
            argumentnew = strrep(argumentnew, "}""", "}");
            argumentnew = strrep(argumentnew, "]""", "]");
            currentresult = cstrcat(currentresult, argumentnew);
            continue;
        end
        if (~ischar(argument) && ~ischar(argument))
            error("Non string argument must be wrapped in {}");
            continue;
        end
        currentresult = cstrcat(currentresult, argument);
    end
    disp(currentresult);
end

function arrayeveryv1result = arrayeveryv1(callbackfunction, anarray)
    % JavaScript-like Array.every() function
    isconditionmatch = false;
    for arrayitemindex = (1:1:numel(anarray)) % (start:step:stop)
        arrayitem = anarray{arrayitemindex};
        isconditionmatch = callbackfunction(arrayitem, arrayitemindex, anarray);
        if (isconditionmatch == false)
            break;
        end
    end
    arrayeveryv1result = isconditionmatch;
end

function arrayeveryv2result = arrayeveryv2(callbackfunction, anarray)
    % JavaScript-like Array.every() function
    isconditionmatch = false;
    for arrayitemindex = (1:1:numel(anarray)) % (start:step:stop)
        arrayitem = anarray{arrayitemindex};
        isconditionmatch = callbackfunction(arrayitem, arrayitemindex, anarray);
        if (isconditionmatch == false)
            arrayeveryv2result = isconditionmatch;
            return;
        end
    end
    arrayeveryv2result = isconditionmatch;
end

function arrayeveryv3result = arrayeveryv3(callbackfunction, anarray)
    % JavaScript-like Array.every() function
    for arrayitemindex = (1:1:numel(anarray)) % (start:step:stop)
        arrayitem = anarray{arrayitemindex};
        isconditionmatch = callbackfunction(arrayitem, arrayitemindex, anarray);
        if (isconditionmatch == false)
            arrayeveryv3result = false;
            return;
        end
    end
    arrayeveryv3result = true;
end

function arrayeveryv4result = arrayeveryv4(callbackfunction, anarray)
    % JavaScript-like Array.every() function
    for arrayitemindex = (1:1:numel(anarray)) % (start:step:stop)
        arrayitem = anarray{arrayitemindex};
        if (callbackfunction(arrayitem, arrayitemindex, anarray) == false)
            arrayeveryv4result = false;
            return;
        end
    end
    arrayeveryv4result = true;
end

sprint(sprintf("\n"), "% JavaScript-like Array.every() in GNU Octave Cell-Array");

numbers = {12, 34, 27, 23, 65, 93, 36, 87, 4, 254};
sprint("numbers: ", jsonstringify(numbers));

sprint("% using JavaScript-like Array.every() function ""arrayeveryv1""");

isallnumberlessthan500 = arrayeveryv1(@(number, varargin) (number < 500), numbers);
sprint("is all number < 500: ", {isallnumberlessthan500});
% is all number < 500: true

isallnumbermorethan500 = arrayeveryv1(@(number, varargin) (number > 500), numbers);
sprint("is all number > 500: ", {isallnumbermorethan500});
% is all number > 500: false

sprint("% using JavaScript-like Array.every() function ""arrayeveryv2""");

isallnumberlessthan500 = arrayeveryv2(@(number, varargin) (number < 500), numbers);
sprint("is all number < 500: ", {isallnumberlessthan500});
% is all number < 500: true

isallnumbermorethan500 = arrayeveryv2(@(number, varargin) (number > 500), numbers);
sprint("is all number > 500: ", {isallnumbermorethan500});
% is all number > 500: false

sprint("% using JavaScript-like Array.every() function ""arrayeveryv3""");

isallnumberlessthan500 = arrayeveryv3(@(number, varargin) (number < 500), numbers);
sprint("is all number < 500: ", {isallnumberlessthan500});
% is all number < 500: true

isallnumbermorethan500 = arrayeveryv3(@(number, varargin) (number > 500), numbers);
sprint("is all number > 500: ", {isallnumbermorethan500});
% is all number > 500: false

sprint("% using JavaScript-like Array.every() function ""arrayeveryv4""");

isallnumberlessthan500 = arrayeveryv4(@(number, varargin) (number < 500), numbers);
sprint("is all number < 500: ", {isallnumberlessthan500});
% is all number < 500: true

isallnumbermorethan500 = arrayeveryv4(@(number, varargin) (number > 500), numbers);
sprint("is all number > 500: ", {isallnumbermorethan500});
% is all number > 500: false

sprint("% using GNU Octave Array.every() built-in function ""all"" and ""cellfun""");

isallnumberlessthan500 = all(cellfun(@(number) (number < 500), numbers));
sprint("is all number < 500: ", {isallnumberlessthan500});
% is all number < 500: true

isallnumbermorethan500 = all(cellfun(@(number) (number > 500), numbers));
sprint("is all number > 500: ", {isallnumbermorethan500});
% is all number > 500: false

sprint(sprintf("\n"), "% JavaScript-like Array.every() in GNU Octave Cell-Array of structs");

products = { ...
    struct( ...
        "code", {"pasta"}, ...
        "price", {321} ...
    ), ...
    struct( ...
        "code", {"bubble_gum"}, ...
        "price", {233} ...
    ), ...
    struct( ...
        "code", {"potato_chips"}, ...
        "price", {5} ...
    ), ...
    struct( ...
        "code", {"towel"}, ...
        "price", {499} ...
    ) ...
};
sprint("products: ", jsonstringify(products, struct("pretty", {true})));

sprint("% using JavaScript-like Array.every() function ""arrayeveryv1""");

isallproductpricelessthan500 = arrayeveryv1(@(product, varargin) (product.("price") < 500), products);
sprint("is all product price < 500: ", {isallproductpricelessthan500});
% is all product price < 500: true

isallproductpricemorethan500 = arrayeveryv1(@(product, varargin) (product.("price") > 500), products);
sprint("is all product price > 500: ", {isallproductpricemorethan500});
% is all product price > 500: false

sprint("% using JavaScript-like Array.every() function ""arrayeveryv2""");

isallproductpricelessthan500 = arrayeveryv2(@(product, varargin) (product.("price") < 500), products);
sprint("is all product price < 500: ", {isallproductpricelessthan500});
% is all product price < 500: true

isallproductpricemorethan500 = arrayeveryv2(@(product, varargin) (product.("price") > 500), products);
sprint("is all product price > 500: ", {isallproductpricemorethan500});
% is all product price > 500: false

sprint("% using JavaScript-like Array.every() function ""arrayeveryv3""");

isallproductpricelessthan500 = arrayeveryv3(@(product, varargin) (product.("price") < 500), products);
sprint("is all product price < 500: ", {isallproductpricelessthan500});
% is all product price < 500: true

isallproductpricemorethan500 = arrayeveryv3(@(product, varargin) (product.("price") > 500), products);
sprint("is all product price > 500: ", {isallproductpricemorethan500});
% is all product price > 500: false

sprint("% using JavaScript-like Array.every() function ""arrayeveryv4""");

isallproductpricelessthan500 = arrayeveryv4(@(product, varargin) (product.("price") < 500), products);
sprint("is all product price < 500: ", {isallproductpricelessthan500});
% is all product price < 500: true

isallproductpricemorethan500 = arrayeveryv4(@(product, varargin) (product.("price") > 500), products);
sprint("is all product price > 500: ", {isallproductpricemorethan500});
% is all product price > 500: false

sprint("% using GNU Octave Array.every() built-in function ""all"" and ""cellfun""");

isallproductpricelessthan500 = all(cellfun(@(product) (product.("price") < 500), products));
sprint("is all product price < 500: ", {isallproductpricelessthan500});
% is all product price < 500: true

isallproductpricemorethan500 = all(cellfun(@(product) (product.("price") > 500), products));
sprint("is all product price > 500: ", {isallproductpricemorethan500});
% is all product price > 500: false
